### PR TITLE
Extract class from CSP configuration/initialization

### PR DESCRIPTION
**Pull request by mjankowski**
As discussed here - https://github.com/mastodon/mastodon/pull/26889#issuecomment-1715762574

This is just a first pass which attempts to do nothing but a) extract a usable class from the CSP config, b) add some coverage around current behavior.

I have attempted to not change any behavior here. Assuming this direction is ok, there's probably more to add here in the way of edge case spec coverage, further refactors to the main initializer file, and then as change behavior or chase down more edge case bugs, expand this coverage when doing so.

### CODE REVIEW

**Comment by renchap** [comment]
Could you update it to cover the changes in #26889 once this is merged?

**Comment by mjankowski** [comment]
> Could you update it to cover the changes in #26889 once this is merged?

Sure. Rebased in a way that incorporates what I think was happening there. It'd be useful to get a spec or other description of the cases that change was handling so I can be more sure.

File: app/lib/content_security_policy.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    end.to_s
+  end
+
+  def host_protocol
+    Rails.configuration.x.use_https ? 'https' : 'http'
+  end
+
+  def host_from_string_with_optional_path(host_string)
+    host_string.split('/').first
+  end
```
Comment (ClearlyClaire) [inline app/lib/content_security_policy.rb:62]: We walked back from splitting the host string in #26889, as paths are valid (and match sub-paths as long as they end with a `/`) in source lists and allow for stricter directives, which can make a meaningful difference with some configurations.

So I think this should be updated accordingly, removing `host_from_string_with_optional_path` and its only use.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/lib/content_security_policy.rb (hunk 1 above)
Comment (mjankowski) [reply app/lib/content_security_policy.rb]: Pushed commit which removes the host string split, but keeps the trailing `/` append.

Let me know if there are other good test values to mix in here based on any of the bug reports. I think this is currently preserving the behavior, but I'd love to have more specs proving that.

File: spec/lib/content_security_policy_spec.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  describe '#media_host' do
+    context 'when there is no configured CDN' do
+      it 'defaults to using the assets_host value' do
+        expect(subject.media_host).to eq(subject.assets_host)
+      end
+    end
+
+    context 'when an S3 alias host is configured' do
+      around do |example|
+        ClimateControl.modify S3_ALIAS_HOST: 'asset-host.s3-alias.example' do
```
Comment (ClearlyClaire) [inline spec/lib/content_security_policy_spec.rb:71]: I think a good test would be something like `S3_ALIAS_HOST: 'files.s3-provider.example/mastodon'`, and check that the `media_host` in this case would be `https://files.s3-provider.example/mastodon/`

Comment (mjankowski) [reply spec/lib/content_security_policy_spec.rb:71]: Pushed spec along those lines.

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**